File: chapter2/2.05/source.c
/*
 * =====================================================================================
 *
 *       Filename:  source.c
 *
 *    Description:  寻找最大的K个数
 *
 *        Version:  1.0
 *        Created:  2012/09/18 17时11分40秒
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  wengsht, wengsht.sysu@gmail.com
 *   Organization:  SYSU - Lab501
 *
 * =====================================================================================
 */
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <string.h>

void print(int *arr, int m, int n);
void swap(int *arr, int i, int j);

/* 
 * ===  FUNCTION  ======================================================================
 *         Name:  solution1
 *  Description:  部分快速排序, 调用将最大的k个数放在arr前k位 最坏情况O(n*ln(n)) 一般可以达到O(n)
 * =====================================================================================
 */
void solution1(int *arr, int k, int n)
{
    if(k <= 0 || k >= n || !arr) return ;
    int i = 0, p = 0;

    int j;
    for(j = 1;j < n;j++)
    {
        if(arr[j] >= arr[i]) swap(arr, ++p, j);
    }
    swap(arr, i, p);

    if(p+1 == k) return ;
    if(p+1 < k)
        solution1(arr+p+1, k-p-1, n-p-1);
    else
        solution1(arr, k, p);
}

/* 
 * ===  FUNCTION  ======================================================================
 *         Name:  solution2
 *  Description:  二分查找第k大数,调用将最大的k个数放在arr前k位,O(ln(MAX-MIN)*n)
 * =====================================================================================
 */
void solution2(int *arr, int k, int n)
{
    if(k <= 0 || k >= n || !arr) return ;

    int mn = arr[0], mx = arr[0];
    int i, c1, c2, mid;
    for(i = 1;i < n; i += 1)
    {
        if(arr[i] > mx) mx = arr[i];
        if(arr[i] < mn) mn = arr[i];
    }

    while(mn <= mx)
    {
        mid = mn + ((mx-mn)>>1);

        c1 = 0;
        c2 = 1;
        for(i = 0; i < n; i += 1)
        {
            if(arr[i] >= mid)
            {
                c1 += 1;
                if(arr[i] > mid)
                    c2 += 1;
            }
        }
        if(c1 >= k && c2 <= k)
            break;

        if(c2 > k) mn = mid+1;
        else       mx = mid-1;
    }

    int p = 0;
    c2--;
    for(i = 0; i < n; i += 1)
        if(arr[i] > mid && c2-- || arr[i] == mid && k-p > c2) swap(arr, i, p++);
}

/* 
 * ===  FUNCTION  ======================================================================
 *         Name:  solution3
 *  Description:  最小优先队列方法，O(n*lnk)，结果存于res中，适用于k大小适中的情况
 * =====================================================================================
 */

void fix_delete(int *heap, int heap_size);
void fix_insert(int *heap, int heap_size);
void solution3(int *arr, int k, int n, int *heap)
{
    if(!arr) return ;
    if(k <= 0 || k >= n)
    {
        memcpy(heap, arr, n*sizeof(int));
        return ;
    }
    int heap_size = 0;
    int i;

    for(i = 0; i < n; i += 1)
    {
        if(heap_size < k)
        {
            heap[++heap_size] = arr[i];
            fix_insert(heap, heap_size);
        }
        else if(arr[i] > heap[1])
        {
            heap[1] = arr[i];
            fix_delete(heap, heap_size);
        }
    }
    swap(heap, 0, heap_size);
}

int main (int argc, char *argv[])
{
    int test[100], tmp[100];
    srand(time(0));

    int i;
    
    for(i = 0; i < 100; i += 1)
        test[i] = rand() % 100;

    printf("原数据:\n");
    print(test, 100, 100);

    memcpy(tmp, test, sizeof(tmp));
    solution1(tmp, 10, 100);
    printf("部分快速排序方法，处理后:\n");
    print(tmp, 10, 100);

    memcpy(tmp, test, sizeof(tmp));
    solution2(tmp, 10, 100);
    printf("二分查找第K大数方法，处理后:\n");
    print(tmp, 10, 100);
    
    solution3(test, 10, 100, tmp);
    printf("最小优先队列方法，处理后:\n");
    print(tmp, 10, 10);
    
    return 0;
}

void print(int *arr, int m, int n)
{
    int i;
    for(i = 0; i < m; i += 1)
        printf("%d ", arr[i]);
    printf(" |  ");
    for(i = m; i < n; i += 1)
        printf("%d ", arr[i]);

    printf("\n");
}

void swap(int *arr, int i, int j)
{
    int tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
}
void fix_delete(int *heap, int heap_size)
{
    int l, r, t;
    int p = 1;
    while(p <= heap_size)
    {
        t = l = p<<1;
        r = l+1;

        if(l <= heap_size && heap[p] > heap[l] || r <= heap_size && heap[p] > heap[r])
        {
            if(r <= heap_size && heap[r] < heap[l]) t = r;

            swap(heap, p, t);

            p = t;
        }
        else break;
    }
}
void fix_insert(int *heap, int heap_size)
{
    int p = heap_size;
    int fa;
    while(p > 1)
    {
        fa = p>>1;
        if(heap[fa] > heap[p])
        {
            swap(heap, fa, p);

            p = fa;
        }
        else break;
    }
}

File: chapter2/2.13/source.py
# -*- coding: utf-8 -*-
####################################################
#
#   2.13 子数组的最大乘积
#    
####################################################

from random import randint

MAX = 10000

# O(2*N) 线性计算从头扫到尾 再扫回来
def solution1(arr, n):
    if n <= 1:
        return 0
    tmp = [arr[0]]
    for i in range(1, n-1):
        tmp.append(tmp[i-1] * arr[i])

    tmp.append(arr[n-1])
    res = tmp[n-2]

    for i in range(n-2, 0, -1):
        res_tmp = tmp[i-1] * tmp[i+1]
        if res_tmp > res:
            res = res_tmp

        tmp[i] = tmp[i+1] * arr[i]

    if tmp[1] > res:
        res = tmp[1]
    return res

# O(N)   利用正负零判断优化速度
def solution2(arr, n):
    res = 0

    p = 1
    q = 1

    # 绝对值最大负数 绝对值最小负数 绝对值最小正数
    a = b = c = 0

    # 0 和 正数个数
    zero = posi = 0
    
    for i in range(0, n):
        if arr[i] != 0:
            p *= arr[i]
        if arr[i] == 0:
            zero += 1
        if arr[i] > 0:
            posi += 1

        if arr[c] <= 0 or arr[i] > 0 and arr[i] < arr[c]:
            c = i

        if arr[b] >= 0 or arr[i] < 0  and arr[i] > arr[b]:
            b = i

        if arr[a] >= 0 or arr[i] < 0 and arr[i] < arr[a]:
            a = i

    if zero >= 2:
        return 0
    if zero == 1:
        return max(0, p)

    m = a
    if p < 0:
        m = b
    else:
        if posi > 0:
            m = c
        else:
            m = a

    res = 1
    for i in range(0, m):
        res *= arr[i]
    for i in range(m+1, n):
        res *= arr[i]

    return res
def main():
    import datetime
    global MAX
    test = []

    # 减少生成0的机会
    for i in range(0, MAX-MAX/100):
        test.append(randint(-100,100))
        while test[i] == 0:
            test[i] = randint(-100, 100)

    for i in range(MAX-MAX/100, MAX):
        test.append(randint(-100,100))

    st = datetime.datetime.now()
    res = solution1(test, MAX)
    ed = datetime.datetime.now()

    print '最大乘积：%d' % res

    t1 = ed - st

    st = datetime.datetime.now()
    res = solution2(test, MAX)
    ed = datetime.datetime.now()

    print '最大乘积：%d' % res

    print '第一种方法耗时：', t1
    print '第二种方法耗时：', ed-st


if __name__ == '__main__':
    main();
